indico/modules/events/registration/apple_pass.py

```diff
@@ -0,0 +1,133 @@
+# This file is part of Indico.
+# Copyright (C) 2002 - 2024 CERN
+#
+# Indico is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see the
+# LICENSE file for more details.
+import json
+import os
+import tempfile
+from pathlib import Path
+
+from cryptography import x509
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.serialization import pkcs7
+from flask import current_app
+from wallet.models import Barcode, BarcodeFormat, EventTicket, Pass
+
+from indico.core import signals
+from indico.core.config import config
+from indico.core.logger import Logger
+from indico.web.flask.util import url_for
+
+
+logger = Logger.get('events.registration.apple_pass')
+
+
+class IndicoPass(Pass):
+
+    def _createSignature(self, manifest, certificate, key, wwdr_certificate, password):  # noqa: N802
+        """Creates a signature (DER encoded) of the manifest.
+
+        Rewritten to use cryptography library instead of M2Crypto
+        The manifest is the file containing a list of files included in the pass file (and their hashes).
+        """
+        # if the certificate has already been converted to Byte, just use it as is.
+        cert = x509.load_pem_x509_certificate(certificate.encode('UTF-8')) if isinstance(certificate, str) else (
+            certificate)
+        private_key = serialization.load_pem_private_key(key.encode('UTF-8'), password=password.encode('UTF-8'))
+        wwdr_cert = x509.load_pem_x509_certificate(Path(wwdr_certificate).read_bytes())
+        options = [pkcs7.PKCS7Options.DetachedSignature]
+        return (
+            pkcs7.PKCS7SignatureBuilder()
+            .set_data(manifest)
+            .add_signer(cert, private_key, hashes.SHA256())
+            .add_certificate(wwdr_cert)
+            .sign(serialization.Encoding.DER, options)
+        )
+
+
+class ApplePassManager:
+    """Apple Pass ticketing manager.
+
+    This class handles the integration with the Apple Pass ticketing API.
+    """
+
+    def __init__(self, event):
+        self.event = event
+        self.settings = self.event.category.effective_apple_pass_config if not self.event.is_unlisted else None
+        self.cert = None
+        self.static_path = os.path.join(current_app.root_path, 'web', 'static')
+
+    @property
+    def configured(self):
+        return config.ENABLE_APPLE_PASS and self.settings is not None
+
+    def build_ticket_object(self, registration):
+        from indico.modules.categories.controllers.util import make_format_event_date_func
```

Comment:
yes, that needs to be local since having an import dependency here would be super ugly (and IIRC when testing the GWM i did have some circular import issues)